dask/dataframe/partitionquantiles.py

```diff
@@ -0,0 +1,397 @@
+
+
+def prepare_percentile_merge(qs, vals, length):
+    """Weigh percentile values by length and the difference between percentiles
+
+    >>> percentiles = np.array([0, 25, 50, 90, 100])
+    >>> values = np.array([2, 3, 5, 8, 13])
+    >>> length = 10
+    >>> prepare_percentile_merge(percentiles, values, length)
+    ([2, 3, 5, 8, 13], [125.0, 250.0, 325.0, 250.0, 50.0])
+
+    The weight of the first element, ``2``, is determined by the difference
+    between the first and second percentiles, and then scaled by length:
+
+    >>> 0.5 * length * (percentiles[1] - percentiles[0])
+    125.0
+
+    The second weight uses the difference of percentiles on both sides, so
+    it will be twice the first weight if the percentiles are equally spaced:
+
+    >>> 0.5 * length * (percentiles[2] - percentiles[0])
+    250.0
+    """
+    if length == 0:
+        return ()
+    diff = np.ediff1d(qs, 0.0, 0.0)
+    weights = 0.5 * length * (diff[1:] + diff[:-1])
+    return vals.tolist(), weights.tolist()
+
+
+def _merge_sorted(items):
+    """Merge and sort items that are already sorted.
+
+    Each item is a tuple like ``(vals, weights)`` where vals and weights
+    are lists.  We sort both by vals.
+    """
+    items = [x for x in items if x]
+    if not items:
+        return ()
+    elif len(items) == 1:
+        return items[0]
+    return tuple(zip(*merge_sorted(*[zip(x, y) for x, y in items])))
+
+
+def process_val_weights(vals_and_weights, finalq, dtype_info):
+    """Calculate final approximate percentiles given weighted vals
+
+    ``vals_and_weights`` is assumed to be sorted.  We take a cumulative
+    sum of the weights, which makes them percentile-like (their scale is
+    [0, N] instead of [0, 100]).  Next we find the divisions that are
+    closest to the desired quantiles, ``finalq``.
+
+    It is possible for adjacent values of the result to be the same.  Since
+    these determine the divisions of the new partitions, some partitions
+    may be empty.
+    """
+    vals, weights = vals_and_weights
+    vals = np.array(vals)
+    weights = np.array(weights)
+
+    # percentile-like, but scaled by weights
+    q = np.cumsum(weights)
+
+    # rescale finalq percentiles to match q
+    desired_q = finalq * q[-1]
+
+    dtype, info = dtype_info
+    if str(dtype) != 'category' and np.issubdtype(vals.dtype, np.number):
+        rv = np.interp(desired_q, q, vals)
+    else:
+        left = np.searchsorted(q, desired_q, side='left')
+        right = np.searchsorted(q, desired_q, side='right') - 1
+        # stay inbounds
+        np.maximum(right, 0, right)
+        lower = np.minimum(left, right)
+        rv = vals[lower]
+
+    if str(dtype) == 'category':
+        rv = pd.Categorical.from_codes(rv, info[0], info[1])
+    elif rv.dtype != dtype:
+        rv = rv.astype(dtype)
+    # Should we try to fudge the result to avoid duplicates?  After all,
+    # two smaller partitions is probably better than having an empty one.
+    return rv
```

Comment:
So, do you think we should fudge the final results to avoid duplicates and empty new partitions?  I don't have a sense of how significant or common this may be, but we can certainly take steps in this function to avoid empty partitions.